akka-actor-typed/src/main/scala/akka/actor/typed/javadsl/ReceiveBuilder.scala

```diff
@@ -33,8 +30,14 @@ final class ReceiveBuilder[T] private (
     val builtSignalHandlers =
       if (signalHandlers.isEmpty) defaultSignalHandlers[T]
       else (adapterExceptionSignalHandler[T] :: signalHandlers).reverse
-
-    new BuiltReceive[T](messageHandlers.reverse, builtSignalHandlers)
+    val hint = messageHandlers.forall(_.`type`.isDefined) && messageHandlers.size >= 5
+    if (hint) {
+      val classifiedMessageHandlers: Map[Class[_ <: T], Array[Case[T, T]]] =
+        messageHandlers.groupBy(_.`type`.get).map { case (clazz, list) => (clazz, list.toArray) }.toMap
```

Comment:
This is not safe, consider the case `onMessage(Object.class, ...).onMessage(String.class, ...)`.

I think it could only ever be done if it can be proven that none of the `case.type` is a supertype of any other, maybe there are more complications as well. 

I think it would be better to separate the simple optimization of using an array and this way more complicated and to be realistic, maybe not possible optimization into separate PRs

Reply:
Thanks, I missed that part. So Let me separate it into two part. The later optimization do need more thought.

Reply:
I think this can be implemented with disjoint set, will try later.